sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamEnumerableConverter.java

```diff
@@ -220,16 +219,18 @@ private static void runCollector(PipelineOptions options, BeamRelNode node) {
     return values;
   }
 
-  private static Enumerable<Object> limitCollect(PipelineOptions options, BeamRelNode node) {
-    long id = options.getOptionsId();
-    ConcurrentLinkedQueue<Row> values = new ConcurrentLinkedQueue<>();
-
+  private static void checkAndFailOnNonDirectRunner(PipelineOptions options, String errorMessage) {
     checkArgument(
         options
             .getRunner()
             .getCanonicalName()
             .equals("org.apache.beam.runners.direct.DirectRunner"),
-        "SELECT without INSERT is only supported in DirectRunner in SQL Shell.");
+        errorMessage);
+  }
+
+  private static Queue<Row> limitCollectRows(PipelineOptions options, BeamRelNode node) {
```

Comment:
I think the queue is to guarantee order of rows for cases where ORDER BY is used?

Reply:
I think there is an order to maintain. Think about this case:

SELECT * FROM a ORDER BY a.col_a DESC LIMIT 5. 

The result should be displayed in SQL shell with a order, which means our collected rows are in a queue (or is a list), rather than a bag of objects.

Reply:
That is true for this case. However the fact that the results internally are ordered is just an implementation detail. Nothing stops us from using queues only for orderby-limit case but use multisets, trees, or anything else instead for other classes of queries. The `toEnumerable()` API doesn't provide the ordering guarantees in general, it's defined at a different level.

Reply:
I can try to say if `Collection` interface will not affect implementation order. If it does maintain the order, I can switch to `Collection` and interpreter it after.